src/Tools/EvtLib/EventLibraryInterface.cxx

```diff
@@ -0,0 +1,309 @@
+#include "Framework/Numerical/RandomGen.h"
+#include "Framework/GHEP/GHepRecord.h"
+#include "Framework/GHEP/GHepParticle.h"
+#include "Framework/ParticleData/PDGUtils.h"
+#include "Framework/ParticleData/PDGLibrary.h"
+#include "Framework/ParticleData/PDGCodes.h"
+#include "Framework/Interaction/Interaction.h"
+#include "Tools/EvtLib/EventLibraryInterface.h"
+#include "Tools/EvtLib/EvtLibRecordList.h"
+#include "Tools/EvtLib/Utils.h"
+
+#include "TFile.h"
+
+using namespace genie;
+using namespace genie::evtlib;
+
+//____________________________________________________________________________
+EventLibraryInterface::EventLibraryInterface() :
+  EventRecordVisitorI("genie::evtlib::EventLibraryInterface"),
+  fRecordFile(0)
+{
+
+}
+
+//____________________________________________________________________________
+EventLibraryInterface::EventLibraryInterface(string config) :
+  EventRecordVisitorI("genie::evtlib::EventLibraryInterface", config),
+  fRecordFile(0)
+{
+
+}
+
+//____________________________________________________________________________
+EventLibraryInterface::~EventLibraryInterface()
+{
+  Cleanup();
+}
+
+//____________________________________________________________________________
+void EventLibraryInterface::ProcessEventRecord(GHepRecord * event) const
+{
+// Get event summary constructed by GENIE
+//
+  Interaction* interaction = event->Summary();
+  const InitialState & init_state = interaction->InitState();
+
+  const EvtLibRecord* rec = GetRecord(interaction);
+  if(!rec) return; // Reason has already been printed
+
+  std::unique_ptr<TLorentzVector> probe_p4(init_state.GetProbeP4(kRfLab));
+
+  unsigned int nLep = 0;
+  for(const EvtLibParticle& part: rec->parts) if(pdg::IsLepton(part.pdg)) ++nLep;
+
+  const int firstLep = (nLep == 0) ? -1 : 2;
+  const int lastLep  = (nLep == 0) ? -1 : 1+nLep;
+
+  const int firstHad = (rec->parts.size() == nLep) ? -1 : 2+nLep;
+  const int lastHad  = (rec->parts.size() == nLep) ? -1 : 1+rec->parts.size();
+
+  // Neutrino is a parent to the lepton(s)
+  event->AddParticle(init_state.ProbePdg(),
+                     kIStInitialState,
+                     -1, -1, firstLep, lastLep,
+                     *probe_p4,
+                     TLorentzVector(0, 0, 0, 0));
+
+  const int tgt_A    = init_state.Tgt().A();
+  const int tgt_Z    = init_state.Tgt().Z();
+  const int tgt_pdgc = pdg::IonPdgCode(tgt_A, tgt_Z);
+  // REVIEW: it was like this in Costas' example code. Can it just be
+  // init_state.TgtPdg()?
+
+  LOG("ELI", pINFO)
+    << "Adding nucleus [A = " << tgt_A << ", Z = " << tgt_Z
+    << ", pdg = " << tgt_pdgc << "]";
+
+  // Nucleus is a parent to everything else
+  event->AddParticle(tgt_pdgc,
+                     kIStInitialState,
+                     -1, -1, firstHad, lastHad,
+                     0, 0, 0, PDGLibrary::Instance()->Find(tgt_pdgc)->Mass(),
+                     0, 0, 0, 0);
+
+  // Include nuclear target in this position (2) by convention
```

Comment:
This is not always the case, it should be dropped and it certainly it is not supposed to be taken from the Summary at this stage